Firmware/HomeButtonsArduino/src/app.cpp

```diff
@@ -0,0 +1,934 @@
+  uint32_t leds_free = uxTaskGetStackHighWaterMark(leds_task_h_);
+  uint32_t main_free = uxTaskGetStackHighWaterMark(main_task_h_);
+  uint32_t num_tasks = uxTaskGetNumberOfTasks();
+  info(
+      "free stack: btns %d, disp %d, net %d, leds %d, main "
+      "%d, num tasks %d",
+      btns_free, disp_free, net_free, leds_free, main_free, num_tasks);
+  uint32_t esp_free_heap = ESP.getFreeHeap();
+  uint32_t esp_min_free_heap = ESP.getMinFreeHeap();
+  uint32_t rtos_free_heap = xPortGetFreeHeapSize();
+  info("free heap: esp %d, esp min %d, rtos %d", esp_free_heap,
+       esp_min_free_heap, rtos_free_heap);
+}
+
+void App::_begin_buttons() {
+  std::pair<uint8_t, uint16_t> button_map[NUM_BUTTONS] = {
+      {hw_.BTN1_PIN, 1}, {hw_.BTN6_PIN, 2}, {hw_.BTN2_PIN, 3},
+      {hw_.BTN5_PIN, 4}, {hw_.BTN3_PIN, 5}, {hw_.BTN4_PIN, 6}};
+  for (uint i = 0; i < NUM_BUTTONS; i++) {
+    buttons_[i].begin(button_map[i].first, button_map[i].second);
+  }
+}
+
+void App::_end_buttons() {
+  for (auto& b : buttons_) {
+    b.end();
+  }
+}
+
+void App::_button_task(void* param) {
+  App* app = static_cast<App*>(param);
+  while (true) {
+    for (auto& b : app->buttons_) {
+      b.update();
+    }
+    delay(20);
+  }
+}
+
+void App::_leds_task(void* param) {
+  App* app = static_cast<App*>(param);
+  while (true) {
+    app->leds_.update(app->hw_);
+    delay(100);
+  }
+}
+
+void App::_display_task(void* param) {
+  App* app = static_cast<App*>(param);
+  while (true) {
+    app->display_.update();
+    delay(50);
+  }
+}
+
+void App::_network_task(void* param) {
+  App* app = static_cast<App*>(param);
+  app->network_.setup();
+  while (true) {
+    app->network_.update();
+    delay(10);
+  }
+}
+
+void App::_start_button_task() {
+  if (button_task_h_ != nullptr) return;
+  debug("button task started.");
+  xTaskCreate(_button_task,    // Function that should be called
+              "BUTTON",        // Name of the task (for debugging)
+              2000,            // Stack size (bytes)
+              this,            // Parameter to pass
+              1,               // Task priority
+              &button_task_h_  // Task handle
+  );
+}
+
+void App::_start_display_task() {
+  if (display_task_h_ != nullptr) return;
+  debug("m_display task started.");
+  xTaskCreate(_display_task,    // Function that should be called
+              "DISPLAY",        // Name of the task (for debugging)
+              5000,             // Stack size (bytes)
+              this,             // Parameter to pass
+              1,                // Task priority
+              &display_task_h_  // Task handle
+  );
+}
+
+void App::_start_network_task() {
+  if (network_task_h_ != nullptr) return;
+  debug("network task started.");
+  xTaskCreate(_network_task,    // Function that should be called
+              "NETWORK",        // Name of the task (for debugging)
+              20000,            // Stack size (bytes)
```

Comment:
What is the reason you increased the stack size for network task? In what scenario was 10k too small?

Reply:
There is indeed an explanation: I replaced a lot of dynamic allocation by fixed-size buffers (which are allocated on the stack).
Now, application is using less heap, but more stack, so I had to increase it.
With this new value, the minimum available space on stack (per task) is the following:
```
free stack: btns 1656, disp 3108, net 9432, leds 212, main 16304, num tasks 11
```
Without this increase, program will crash for sure (stack smashing detection mechanism will be triggered).

Note: the log is titled `free stack` but in fact, it is the minimum space available since boot (`uxTaskGetStackHighWaterMark`)

Reply:
Note: do define correct stack size, it should be big enough to fit the worst case (+ some extra margins).
From what I found, the worst case is after first setup (sending auto-discovery).
With the 20k value, I never had any issues.